src/common/commonutils/PackageUtils.c

```diff
@@ -90,51 +105,168 @@ static int CheckOrInstallPackage(const char* commandTemplate, const char* packag
 
     status = ExecuteCommand(NULL, command, false, false, 0, g_packageManagerTimeoutSeconds, NULL, NULL, log);
 
-    OsConfigLogInfo(log, "Package manager '%s' command '%s' returning %d (errno: %d)", packageManager, command, status, errno);
+    OsConfigLogInfo(log, "Package manager '%s' command '%s' returning %d", packageManager, command, status);
 
     FREE_MEMORY(command);
 
+    // Refresh the cache holding the list of installed packages next time we check
+    g_updateInstalledPackagesCache = 1;
+
     return status;
 }
 
-int IsPackageInstalled(const char* packageName, OsConfigLogHandle log)
+static int CheckAllPackages(const char* commandTemplate, const char* packageManager, char** results, OsConfigLogHandle log)
 {
-    const char* commandTemplateDpkg = "%s -l %s | grep ^ii";
-    const char* commandTemplateYumDnf = "%s list installed  --cacheonly %s";
-    const char* commandTemplateRedHat = "%s list installed %s --disableplugin subscription-manager";
-    const char* commandTemplateZypper = "%s se -x %s";
+    char* command = NULL;
+    int status = ENOENT;
+
+    if ((NULL == commandTemplate) || (NULL == packageManager) || (NULL == results))
+    {
+        OsConfigLogError(log, "CheckAllPackages called with invalid arguments");
+        return EINVAL;
+    }
+
+    if (NULL == (command = FormatAllocateString(commandTemplate, packageManager)))
+    {
+        OsConfigLogError(log, "CheckAllPackages: FormatAllocateString failed");
+        return ENOMEM;
+    }
+
+    status = ExecuteCommand(NULL, command, false, false, 0, g_packageManagerTimeoutSeconds, results, NULL, log);
+
+    OsConfigLogInfo(log, "Package manager '%s' command '%s' returning  %d", packageManager, command, status);
+    OsConfigLogDebug(log, "%s", *results);
+
+    FREE_MEMORY(command);
+
+    return status;
+}
+
+static int UpdateInstalledPackagesCache(OsConfigLogHandle log)
+{
+    const char* commandTemplateDpkg = "%s-query -W -f='${binary:Package}\n'";
+    const char* commandTemplateYumDnf = "%s list installed  --cacheonly | awk '{print $1}'";
+    const char* commandTmeplateZypper = "%s search -i";
```

Comment:
This might be better kept for a follow-up but someone suggested using `rpm -qa --queryformat "%{NAME}\n"` for all rpm distros, which produces the same output and skips locking/refresh.

Reply:
Awesome! I will try this.

Reply:
Rpm query works great and is more efficient than going the tdnf/yum/zypper route. We keep the later to try for those special cases when neither dpkg or rpm are present, but another package manager may be present. 

It's also better for the Hot Fix in particular, as with rpm we will put even less strain on zypper (what the customer complained about).